packages/marshal/src/marshal.js

```diff
@@ -142,24 +147,90 @@ export const makeMarshal = (
       }
     };
 
-    const encodeRemotableToCapData = val => {
-      const iface = getInterfaceOf(val);
-      // console.log(`serializeSlot: ${val}`);
-      return serializeSlot(val, iface);
-    };
+    if (useSmallcaps) {
+      /**
+       * @param {string} prefix
+       * @param {Passable} passable
+       * @param {string} [iface]
+       */
+      const serializeSlotToSmallcaps = (
+        prefix,
+        passable,
+        iface = undefined,
+      ) => {
+        let slotIndex;
+        if (slotMap.has(passable)) {
+          // TODO assert that it's the same iface as before
+          slotIndex = slotMap.get(passable);
+          assert.typeof(slotIndex, 'number');
+          iface = undefined;
+        } else {
+          const slot = convertValToSlot(passable);
+          slotIndex = slots.length;
+          slots.push(slot);
+          slotMap.set(passable, slotIndex);
+        }
 
-    const encode = makeEncodeToCapData({
-      encodeRemotableToCapData,
-      encodePromiseToCapData: serializeSlot,
-      encodeErrorToCapData,
-    });
+        // TODO explore removing this special case
+        if (iface === undefined) {
+          return `${prefix}${slotIndex}`;
+        }
+        return `${prefix}${slotIndex}.${iface}`;
+      };
 
-    const encoded = encode(root);
+      const encodeRemotableToSmallcaps = remotable => {
+        const iface = getInterfaceOf(remotable);
+        // console.log(`serializeSlot: ${val}`);
+        return serializeSlotToSmallcaps('$', remotable, iface);
+      };
 
-    return harden({
-      body: JSON.stringify(encoded),
-      slots,
-    });
+      const encodePromiseToSmallcaps = promise => {
+        return serializeSlotToSmallcaps('&', promise);
+      };
+
+      const encodeErrorToSmallcaps = err => {
+        // Not the most elegant way to reuse code. TODO refactor.
+        const capDataErr = encodeErrorToCapData(err);
+        const { [QCLASS]: _, message, ...rest } = capDataErr;
+        return harden({
+          '#error': message,
+          ...rest,
```

Comment:
Should `rest` itself be encoded? If it isn't, then an error object with e.g. a "#tag" property will be encoded into an invalid structure.

Reply:
This `rest` is not about the rest of the error object. It is a way for `encodeErrorToSmallcaps` to delegate most of the job to `encodeErrorToCapData`, and then to say (by destructuring and restructuring) "My encoding is just like that encoding except for these differences."

However, thinking about this reveals a bigger problem: the `message`, `rest`, and optional `errorId` property names are fine. Encoding them or not wouldn't make any difference. However, the string value of these properties are not encoded, which is technically "correct" because the decoder doesn't decode them. However, we designed to be more human readable than capData. To someone reading smallcaps, they will habituate to running the decoder in their head. They will do so here as well. Enabling users to provide unencoded strings that might look like encoded strings enables them to mislead readers.

Fixing this should also fix a related fragility: The ability of `encodeErrorToSmallcaps` to delegate some of the encoding to `encodeErrorToCapData` relies on capData *not* encoding these strings to capData, which smallcaps should not be burdened with understanding. Will update soon with a solution to these problems.

Reply:
Done, by two significant changes:
   * The parts of errors (`name`, `message`, and optional `errorId`) now recursively encoded.
   * To enable this, I changed the interface between marshal and its optional plugin encoders/decoders to take the overall recursive encoder/decoder as an argument.